pinot-core/src/main/java/org/apache/pinot/core/data/manager/offline/DimensionTableDataManager.java

```diff
@@ -117,6 +135,26 @@ public void removeSegment(String segmentName) {
     }
   }
 
+  @Override
+  protected void doShutdown() {
+    closeDimensionTable(_dimensionTable, _disablePreload);
+  }
+
+  private void closeDimensionTable(DimensionTable dimensionTable, boolean releaseSegments) {
+    if (dimensionTable != null) {
+      try {
+        dimensionTable.close();
+        if (releaseSegments && _segmentDataManagers != null) {
+          for (SegmentDataManager segmentDataManager: _segmentDataManagers) {
```

Comment:
This doesn't work because the current `_segmentDataManagers` are for the new dimension table, and here we are closing the already replaced dimension table.

We can pass the acquired `List<SegmentDataManager>` into the `MemoryOptimizedDimensionTable`, and close it within the `MemoryOptimizedDimensionTable.close()`. Basically the dimension table implementation knows whether it needs to release the segments, instead of the caller